include/RAJA/pattern/forall.hpp

```diff
@@ -589,6 +589,136 @@ RAJA_INLINE camp::resources::EventProxy<Res> CallForallIcount::operator()(T cons
 
 }  // namespace detail
 
+//
+// Experimental support for dynamic policy selection
+//
+namespace expt
+{
+
+
+
+template<typename T, typename t_pol, typename SEGMENT, typename BODY>
+struct IsHostResource {
+  static const bool value = false;
+
+#if defined(RAJA_DEVICE_ACTIVE)
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, SEGMENT const &seg, BODY const &body)
+  {
+#if defined(RAJA_CUDA_ACTIVE)
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Cuda>(), seg, body);
+#endif
+
+#if defined(RAJA_HIP_ACTIVE)
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Hip>(), seg, body);
+#endif
+
+    return resources::EventProxy<resources::Resource>(r);
+  }
+#endif
+};
+
+template<typename t_pol, typename SEGMENT, typename BODY>
+struct IsHostResource<RAJA::resources::Host, t_pol, SEGMENT, BODY> {
+static const bool value = true;
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, SEGMENT const &seg, BODY const &body)
+  {
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Host>(), seg, body);
+    return resources::EventProxy<resources::Resource>(r);
+  }
+};
+
+template<camp::idx_t IDX, typename POLICY_LIST>
+struct dynamic_helper
+{
+  template<typename SEGMENT, typename BODY>
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, const int pol, SEGMENT const &seg, BODY const &body)
+  {
+
+    using t_pol = typename camp::at<POLICY_LIST,camp::num<IDX>>::type;
+
+    if(IDX==pol){
+      return IsHostResource<typename resources::get_resource<t_pol>::type,
+                            t_pol, SEGMENT, BODY>::invoke_forall(r, seg, body);
+    }
+
+    return dynamic_helper<IDX-1, POLICY_LIST>::invoke_forall(r, pol, seg, body);
+  }
+
+  template<typename SEGMENT, typename BODY>
+  static void invoke_forall(const int pol, SEGMENT const &seg, BODY const &body)
```

Comment:
Should the non-resource function always be written first for consistency?

Reply:
yes!